CO-1107: Wait to Hibernate ClusterPool Clusters until SyncSets Applied

abutcher opened the pull request:
*  Set a `ClusterSync.Status.FirstSuccessTime` of `time.Now()` when there are no `SyncStatuses` found for `ClusterSync`. This means there are no `SyncSets` to apply.
* Determine if `SyncSets` have been applied in hibernation controller
  * If no `ClusterSync.Status.FirstSuccessTime` has been set for 10 minutes since `ClusterDeployment.Status.InstalledTimestamp`, hibernate anyway.
  * If 10 minutes have not passed since install and `ClusterSync.Status.FirstSuccessTime` is unset, requeue `ClusterDeployment` for time remaining to meet 10 minutes since cluster was installed.

/assign @dgoodwin 
/cc @akhil-rane

dgoodwin commented:
Would this make sense on the hibernation controller instead of clusterpool? I'm thinking the cluster pool doesn't really care about this, it's job is just to create the clusters. The hibernation controller when reconciling can just check, are all syncsets applied, and if not are you installed more than 10 minutes ago, and requeue for that 10 minute threshold if not. We then would not need to reconcile a pool every time a clustersync changes for a clusterdeployment from that pool.

We'd essentially be changing from saying it's the job of the clusterpool controller to set the desired state when it knows syncsets are applied, to saying hibernation only kicks in when syncsets are applied or the cluster is over 10 minutes since installed across the board.

WDYT?

abutcher commented:
I think that’s fine if we’re okay with SyncSets driving all hibernation. I
implemented the change in the hibernation controller originally and moved
it all to pool and it was pretty simple on that side.

On Thu, Nov 12, 2020 at 8:56 AM Devan Goodwin <notifications@github.com>
wrote:

> Would this make sense on the hibernation controller instead of
> clusterpool? I'm thinking the cluster pool doesn't really care about this,
> it's job is just to create the clusters. The hibernation controller when
> reconciling can just check, are all syncsets applied, and if not are you
> installed more than 10 minutes ago, and requeue for that 10 minute
> threshold if not. We then would not need to reconcile a pool every time a
> clustersync changes for a clusterdeployment from that pool.
>
> We'd essentially be changing from saying it's the job of the clusterpool
> controller to set the desired state when it knows syncsets are applied, to
> saying hibernation only kicks in when syncsets are applied or the cluster
> is over 10 minutes since installed across the board.
>
> WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/hive/pull/1205#issuecomment-726093127>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALPG6TRA5WLAIOVKOXBC3SPPSQNANCNFSM4TQ637LQ>
> .
>

dgoodwin commented:
Ok cool, lets get some team feedback before making any changes.

akhil-rane commented on pkg/controller/clusterpool/clusterpool_controller.go:
```diff
@@ -234,6 +278,37 @@ func (r *ReconcileClusterPool) Reconcile(request reconcile.Request) (reconcile.R
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileClusterPool) hibernateReadyClusters(cds []*hivev1.ClusterDeployment) ([]*hivev1.ClusterDeployment, error) {
```
This function returns ClusterDeployments for which syncsets have not been applied. Should we name it to suggest that?

akhil-rane commented on pkg/controller/clusterpool/clusterpool_controller.go:
```diff
@@ -26,18 +27,20 @@ import (
+	imageSetDependent                = "cluster image set"
+	pullSecretDependent              = "pull secret"
+	credentialsSecretDependent       = "credentials secret"
+	hibernateAfterSyncSetsNotApplied = time.Minute * 10
```
Some documentation to indicate what this variable is doing will be helpful

akhil-rane commented on pkg/controller/clusterpool/clusterpool_controller.go:
```diff
@@ -26,18 +27,20 @@ import (
+	imageSetDependent                = "cluster image set"
+	pullSecretDependent              = "pull secret"
+	credentialsSecretDependent       = "credentials secret"
+	hibernateAfterSyncSetsNotApplied = time.Minute * 10
```
Some documentation to indicate what this variable is doing will be really helpful

abutcher replied:
This has moved to the hibernation controller but I added a comment for it.

akhil-rane commented on pkg/controller/clusterpool/clusterpool_controller.go:
```diff
@@ -234,6 +278,37 @@ func (r *ReconcileClusterPool) Reconcile(request reconcile.Request) (reconcile.R
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileClusterPool) hibernateReadyClusters(cds []*hivev1.ClusterDeployment) ([]*hivev1.ClusterDeployment, error) {
```
This function returns ClusterDeployments for which syncsets have not been applied. Should we name it to suggest that?

abutcher replied:
This function has gone away in the refactor.

akhil-rane reviewed (commented):
I agree with @dgoodwin 's suggestion here. Its the job of the hibernation controller to ensure that all conditions are met before hibernating a cluster.

akhil-rane commented:
I loved the way you have structured the commit message :)

dgoodwin commented on pkg/apis/hive/v1/clusterdeployment_types.go:
```diff
@@ -360,6 +360,9 @@ const (
 	FailedToStartHibernationReason = "FailedToStart"
+	// SyncSetsNotAppliedReason is used as the reason when SyncSets have not yet been applied
+	// for the cluster based on ClusterSync.Status.FirstSucessTime
+	SyncSetsNotAppliedReason = "SyncSetsNotApplied"
```
Used on the hibernating condition, good call.

dgoodwin commented on pkg/controller/hibernation/hibernation_controller.go:
```diff
@@ -157,16 +163,23 @@ func (r *hibernationReconciler) Reconcile(request reconcile.Request) (result rec
+			// Allow hibernation if hibernateAfterSyncSetsNotApplied duration has passed since cluster installed and syncsets still not applied
+			if cd.Status.InstalledTimestamp != nil && time.Now().Sub(cd.Status.InstalledTimestamp.Time) < hibernateAfterSyncSetsNotApplied {
+				return r.setHibernatingCondition(cd, hivev1.SyncSetsNotAppliedReason, msg, corev1.ConditionFalse, cdLog)
+			}
```
When this is not the case and we've past expiry, is there any indication on the cluster deployment that we were not able to apply syncsets even though hibernating is now true? I suspect there probably is, via our pre-existing syncsets failing condition. If so this is great.

abutcher replied:
We'll have a `SyncSetApplyFailure` failing condition, right.

```  - lastProbeTime: "2020-11-18T15:40:04Z"
    lastTransitionTime: "2020-11-18T15:40:04Z"
    message: One of the SyncSet applies has failed
    reason: SyncSetApplyFailure
    status: "True"
    type: SyncSetFailed
  - lastProbeTime: "2020-11-18T15:40:04Z"
    lastTransitionTime: "2020-11-18T15:40:02Z"
    message: Cluster SyncSets have not been applied
    reason: SyncSetsNotApplied
    status: "False"```

dgoodwin commented on pkg/controller/hibernation/hibernation_controller.go:
```diff
@@ -346,12 +359,23 @@ func (r *hibernationReconciler) setHibernatingCondition(cd *hivev1.ClusterDeploy
+			logger.Infof("cluster will reconcile due to syncsets not applied in: %v", requeueAfter)
+			result.RequeueAfter = requeueAfter
+			result.Requeue = true
+		}()
```
Looked like this might run after the if, but I looked it up and it appears it probably runs after the function we're in, so :+1:

dgoodwin commented on pkg/controller/hibernation/hibernation_controller.go:
```diff
@@ -383,6 +407,18 @@ func (r *hibernationReconciler) canHibernate(cd *hivev1.ClusterDeployment) (bool
+	clusterSync := &hiveintv1alpha1.ClusterSync{}
+	err := r.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Name}, clusterSync)
+	if err != nil {
+		return false, fmt.Sprintf("could not get ClusterSync: %v", err)
```
feels like we may need to return an error here and force a requeue rather than following the path for normal execution when we're waiting for syncsets to reach first success.

abutcher replied:
Thinking about this I'm not sure that we need to error and go into back off if we're requeuing for the time to force hibernate. I could do this though by squashing the returns from setting the condition and returning an err from reconcile.

abutcher replied:
Suppose it would hibernate sooner in the event that the syncsets did succeed.

dgoodwin commented on pkg/controller/hibernation/hibernation_controller_test.go:
```diff
@@ -73,23 +81,51 @@ func TestReconcile(t *testing.T) {
+				assert.Equal(t, corev1.ConditionTrue, cond.Status)
+				assert.Equal(t, hivev1.StoppingHibernationReason, cond.Reason)
+			},
+		},
```
Would it be feasible to add a test for when there are no syncsets to apply or is that covered elsewhere. I guess it's every other test isn't it.

dgoodwin replied:
Also are we missing a test for when SyncSets have been applied successfully?

abutcher replied:
Added a test to ensure `cs.Status.FirstSuccessTime` is set w/ no syncsets and another in hibernation for syncsets applied successfully.

dgoodwin commented:
/lgtm

abutcher commented:
/test e2e
